crypto/store/loader_file.c

```diff
@@ -0,0 +1,1027 @@
+                                        void *ui_data, int *matchcount)
+{
+    STORE_INFO *result = NULL;
+    BUF_MEM *new_mem = NULL;
+    char *new_pem_name = NULL;
+    int t = 0;
+
+ again:
+    {
+        size_t i = 0;
+        void *handler_ctx = NULL;
+        const STORE_FILE_HANDLER **matching_handlers =
+            OPENSSL_zalloc(sizeof(*matching_handlers)
+                           * OSSL_NELEM(file_handlers));
+
+        if (matching_handlers == NULL) {
+            STOREerr(STORE_F_FILE_LOAD_TRY_DECODE, ERR_R_MALLOC_FAILURE);
+            goto err;
+        }
+
+        *matchcount = 0;
+        for (i = 0; i < OSSL_NELEM(file_handlers); i++) {
+            const STORE_FILE_HANDLER *handler = file_handlers[i];
+            void *tmp_handler_ctx = NULL;
+            STORE_INFO *tmp_result = handler->try_decode(pem_name, pem_header,
+                                                         data, len,
+                                                         &tmp_handler_ctx,
+                                                         ui_method, ui_data);
+
+            if (tmp_result == NULL) {
+                STOREerr(STORE_F_FILE_LOAD_TRY_DECODE, STORE_R_IS_NOT_A);
+                ERR_add_error_data(1, handler->name);
+            } else {
+                if (matching_handlers)
+                    matching_handlers[*matchcount] = handler;
+
+                if (handler_ctx)
+                handler->destroy_ctx(&handler_ctx);
+                handler_ctx = tmp_handler_ctx;
+
+                if (++*matchcount == 1) {
+                    result = tmp_result;
+                    tmp_result = NULL;
+                } else {
+                    /* more than one match => ambiguous, kill any result */
+                    STORE_INFO_free(result);
+                    STORE_INFO_free(tmp_result);
+                    if (handler->destroy_ctx != NULL)
+                        handler->destroy_ctx(&handler_ctx);
+                    handler_ctx = NULL;
+                    result = NULL;
+                }
+            }
+        }
+
+        if (*matchcount > 1)
+            STOREerr(STORE_F_FILE_LOAD_TRY_DECODE,
+                     STORE_R_AMBIGUOUS_CONTENT_TYPE);
+        if (*matchcount == 0)
+            STOREerr(STORE_F_FILE_LOAD_TRY_DECODE,
+                     STORE_R_UNSUPPORTED_CONTENT_TYPE);
+        else if (matching_handlers[0]->repeatable) {
+            if (ctx == NULL) {
+                STOREerr(STORE_F_FILE_LOAD_TRY_DECODE,
+                         STORE_R_UNSUPPORTED_CONTENT_TYPE);
+                STORE_INFO_free(result);
+                result = NULL;
+            } else {
+                ctx->_.file.last_handler = matching_handlers[0];
+                ctx->_.file.last_handler_ctx = handler_ctx;
+            }
+        }
+
+        OPENSSL_free(matching_handlers);
+    }
+
+ err:
+    if (new_pem_name != NULL)
+        OPENSSL_free(new_pem_name);
+    if (new_mem != NULL)
+        BUF_MEM_free(new_mem);
```

Comment:
and if you didn't call BUF_MEM_new you can't call BUF_MEM_free because you don't know what other internal stuff it is doing.